From 79cc84326df09cc8ce307ca37d8daa4d37565b56 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Thu, 28 Aug 2014 19:15:58 -0400
Subject: [virt] virtio_blk: use blk_mq_complete_request

Message-id: <1409253371-23161-9-git-send-email-jmoyer@redhat.com>
Patchwork-id: 89851
O-Subject: [RHEL7 PATCH 08/21] virtio_blk: use blk_mq_complete_request
Bugzilla: 1105698
RH-Acked-by: Fam Zheng <famz@redhat.com>

This is a backport of the following upstream commit.
Related to bug 1105698.

  commit 5124c285797aa33d5fdb909fbe808a61bcc5eb9d
  Author: Christoph Hellwig <hch@infradead.org>
  Date:   Mon Feb 10 03:24:39 2014 -0800

    virtio_blk: use blk_mq_complete_request

    Make sure to complete requests on the submitting CPU.  Previously this
    was done in blk_mq_end_io, but the responsibility shifted to the drivers.

    Signed-off-by: Christoph Hellwig <hch@lst.de>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/virtio_blk.c b/drivers/block/virtio_blk.c
index 6a680d4..b1cb3f4 100644
--- a/drivers/block/virtio_blk.c
+++ b/drivers/block/virtio_blk.c
@@ -110,9 +110,9 @@ static int __virtblk_add_req(struct virtqueue *vq,
  return virtqueue_add_sgs(vq, sgs, num_out, num_in, vbr, GFP_ATOMIC);
 }
 
-static inline void virtblk_request_done(struct virtblk_req *vbr)
+static inline void virtblk_request_done(struct request *req)
 {
- struct request *req = vbr->req;
+ struct virtblk_req *vbr = req->special;
  int error = virtblk_result(vbr);
 
  if (req->cmd_type == REQ_TYPE_BLOCK_PC) {
@@ -138,7 +138,7 @@ static void virtblk_done(struct virtqueue *vq)
  do {
   virtqueue_disable_cb(vq);
   while ((vbr = virtqueue_get_buf(vblk->vq, &len)) != NULL) {
-   virtblk_request_done(vbr);
+   blk_mq_complete_request(vbr->req);
    req_done = true;
   }
   if (unlikely(virtqueue_is_broken(vq)))
@@ -479,6 +479,7 @@ static struct blk_mq_ops virtio_mq_ops = {
  .map_queue = blk_mq_map_queue,
  .alloc_hctx = blk_mq_alloc_single_hw_queue,
  .free_hctx = blk_mq_free_single_hw_queue,
+ .complete = virtblk_request_done,
 };
 
 static struct blk_mq_reg virtio_mq_reg = {
-- 
1.7.1